app/scripts/models/auth_brokers/web-channel.js

```diff
@@ -96,14 +96,43 @@ define([
                 this, account, { closeWindow: true });
     },
 
+    beforeSignUpConfirmationPoll: function (account) {
+      // If the relier wants keys, the signup verification tab will need
+      // to be able to fetch them in order to complete the flow.
+      // Send them as part of the oauth session data.
+      if (this.relier.wantsKeys()) {
+        this.session.set('oauth', _.extend({}, this.session.oauth, {
+          keyFetchToken: account.get('keyFetchToken'),
+          unwrapBKey: account.get('unwrapBKey')
+        }));
+      }
+    },
+
+    afterSignUpConfirmationPoll: function (account) {
+      // The original tab can finish the OAuth flow if it is still open,
+      // but not if the verification tab has already finished it.
+      this.session.load();
+      if (this.session.oauth) {
+        return this.finishOAuthFlow(account);
+      }
+    },
+
     afterCompleteSignUp: function (account) {
       // The original tab may be closed, so the verification tab should
       // send the OAuth result to the browser to ensure the flow completes.
-      //
-      // The slight delay is to allow the functional tests time to bind
-      // event handlers before the flow completes.
-      var self = this;
-      return p().delay(100).then(_.bind(self.finishOAuthFlow, self, account));
+      // If the original tab is still open, there's a small chance that
+      // it could complete the flow before we do.
+      this.session.load();
+      if (this.session.oauth) {
+        if (this.relier.wantsKeys()) {
+          account.set('keyFetchToken', this.session.oauth.keyFetchToken);
+          account.set('unwrapBKey', this.session.oauth.unwrapBKey);
+        }
+        // The slight delay is to allow the functional tests time to bind
+        // event handlers before the flow completes.
+        return p().delay(100)
```

Comment:
This leaves open a 100ms window where the original tab could see that `this.session.oauth` exists, then finish the oauth flow, and the verification tab also finishes the oauth flow just afterwards. 

I would wrap all of your new logic inside of:
```js
  var self = this;
  return p().delay(100)
    .then(function () {
      self.session.load(); // or sync!
      if (self.session.oauth) {
      ....
         return self.finishOAuthFlow(account);
      }
  });
```